---
title: "Whole genome sequencing, align reads, call variants"
author: "Rebecca Batstone"
date: "`r format(Sys.Date())`"
output: github_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Checking original sequences with FastQC

We used: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)

# Use seqtk to clean reads (recommended by CBW)

## need to download to server:
```{bash get_seqtk, eval = FALSE}
git clone git://github.com/lh3/seqtk.git
#./seqtk/seqtk
# put in same directory as samples
```

## make a sample list

needed to ls *fastq, but narrow window so each sample appears on a line
copied
pasted into vim sample_list
- need to press i to insert, press ESC, 
- then colon (:) and x to save (or q! to not save), and quit

## using seqtk on all samples

made vim trim_samples.sh

Make a for loop:
```{bash seqtk_run, eval = FALSE}
# make a shell script in vim (trim_samples.sh)
i
--
while read name;
do ./seqtk/seqtk trimfq -q 0.05 "$name" > trim_"$name";

done < sample_list
--
ESC 
:x

# saved, and then ran
nohup bash trim_samples.sh &
  
# make a list of trimmed samples using vim (just like above)
trim_sample_list  
```

# Reference sequence prep

## sequence download, then unzip
```{r get_ref_seq, engine = 'bash', eval = FALSE}
# KH35c
wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/002/197/105/GCA_002197105.1_ASM219710v1/GCA_002197105.1_ASM219710v1_genomic.fna.gz

# unzip
gunzip GCA_002197105.1_ASM219710v1_genomic.fna.gz
```

## index reference using bwa
```{bash bwa_index_ref, eval = FALSE}
bwa index GCA_002197105.1_ASM219710v1_genomic.fna
# creates a bunch of .fna files w/different extensions
```

## extract reference w/ samtools
```{bash samtools_index_ref, eval = FALSE}
samtools faidx GCA_002197105.1_ASM219710v1_genomic.fna
# creates .fai file
```

## make the .dict file using picard
```{bash picard_dict_ref, eval = FALSE}
java -jar picard.jar CreateSequenceDictionary REFERENCE=GCA_002197105.1_ASM219710v1_genomic.fna OUTPUT=GCA_002197105.1_ASM219710v1_genomic.fna.dict
# makes a .dict file
```

# Alignment: GATK best practices workflow

Ended up just wget'ing the original reads, re-doing seqtk. Now everthing is in align_KH35c_08Mar2018

We begin by mapping the sequence reads to the reference genome to produce a file in SAM/BAM format sorted by coordinate. Next, we mark duplicates to mitigate biases introduced by data generation steps such as PCR amplification. Finally, we recalibrate the base quality scores, because the variant calling algorithms rely heavily on the quality scores assigned to the individual base calls in each sequence read.

## Align reads to reference
map reads to reference (mem works better if sequence quality is good)

Note about modifying read names
e.g., trim_A01-1021-11_171006_NextSeq_R1.fastq
n=${name%_Next*} # n becomes trim_A01-1021-11_171006
n=${name%_R*} # n becomes trim_A01-1021-11_171006_NextSeq

```{bash bwa_align, eval = FALSE}
# make a shell script using vim (bwa_mem.sh)
i
--  
while read name;
do bwa mem -t 4 -M -R "@RG\tID:"${name%_171006*}"\tSM:"${name%_171006*}"" GCA_002197105.1_ASM219710v1_genomic.fna "${name%_R*}"_R1.fastq "${name%_R*}"_R2.fastq > "${name%_171006*}".sam;

done < trimmed_samples.list
--
ESC
:x

# run (in background)
nohup bash bwa_mem.sh > bwa_mem.out &
  
# run bwa for WSM1022
nohup bwa mem -t 4 -M -R "@RG\tID:WSM1022\tSM:WSM1022" GCA_002197105.1_ASM219710v1_genomic.fna GCA_000510665.1_ASM51066v1_genomic.fna > WSM1022.sam &
  
# make a list of samples using vim
sam_samples.list  
```

## Convert sam to bam
```{bash sam_to_bam, eval = FALSE}
# make a shell script using vim (sam_to_bam.sh)
i
--
while read name;
do samtools view -huS -o "${name%.sam*}".bam "$name";

done < sam_samples.list # made list same way as previous
--
ESC
:x

# run (in background)
nohup bash sam_to_bam.sh > sam_to_bam.out &
  
# make a list of samples in vim
bam_samples.list
```

## Reorder using PICARD
```{bash reorder, eval = FALSE}
# make a shell script using vim (picard_reorder.sh)
i
--
while read name;
do java -jar picard.jar ReorderSam R=GCA_002197105.1_ASM219710v1_genomic.fna I="$name" O="$name".reorder;

done < bam_samples.list # made list same way as previous
--
ESC
:x

# run (in background)
nohup bash picard_reorder.sh > picard_reorder.out &
  
# make a sample list using vim
reorder_samples.list
```

```{bash adding_readgroups, eval = FALSE}
# make a shell script (add_read_groups.sh), and then bash it
vim add_read_groups.sh
i
--
while read name;
do java -jar picard.jar AddOrReplaceReadGroups I="$name" O="$name".RG RGLB=Rhizo_evo RGPU=1 RGPL=Illumina RGSM="$name" 

done < reorder_samples.list
--
ESC
:x

# run shell script in background
nohup bash add_RG.sh > add_RG.out &
  
# after it runs, make a list of samples w/ read groups added
vim RG_samples.list
PASTE all names
ESC
:x
```

## Sorting using Picard

```{bash sort, eval = FALSE}
# make a shell script using vim (sort_sam.sh)
i
--
while read name;
do java -jar picard.jar SortSam I="$name" O="$name".sorted SO=coordinate;

done < RG_samples.list 
--
ESC
:x

# run (in background)
nohup bash sort_sam.sh > sort_sam.sh &
  
# make a list of samples using vim
sorted_samples.list
```

## Mark duplicates

```{bash duplicates, eval = FALSE}
# make a shell script (markdups.sh)
i
--
while read name;
do java -jar picard.jar MarkDuplicates I="$name" O="$name".markdups M=duplicates;

done < sorted_samples.list 
--

# run (in background)
nohup bash markdups.sh > markdups.out &
  
# make a list of samples using vim (note the period rather than underscore)
markdups_samples.list
```

## Index the bam (after marked dups)
```{bash index, eval = FALSE}
# make a shell script using vim (index.sh)
i
--
while read name;
do java -jar picard.jar BuildBamIndex I="$name";

done < markdups_samples.list
--
  
# run (in background)
nohup bash index.sh > index.out &
```

# GATK variant discovery

## Haplotype caller

```{bash hap_call, eval = FALSE}
# example renaming:
Rename all *.markdups to *.markdups.bam
for f in *.markdups; do 
mv -- "$f" "${f%.markdups}.markdups.bam"
done

# e.g., hap_call_1-5.sh:
while read name;
do $GATK HaplotypeCaller -R GCA_002197105.1_ASM219710v1_genomic.fna -I "$name" --dont-use-soft-clipped-bases TRUE -ploidy 1 -O "$name".g.vcf -ERC GVCF;

done < samples_1-5.list

nohup bash hap_call_1-5.sh > hap_call_1-5.out &
  
# ran all 10 .sh files at a time. Took ~ 3 days
```

## Run Combine GVCFs

```{bash combine_GVCF, eval = FALSE}
nohup $GATK CombineGVCFs -R GCA_002197105.1_ASM219710v1_genomic.fna -V gVCFs.list -O comb_KH35c.vcf > combine_gVCFs.out &
```

## GenotypeGVCFs 

```{bash genotypeGVCF, eval = FALSE}
nohup $GATK GenotypeGVCFs -R GCA_002197105.1_ASM219710v1_genomic.fna -V comb_KH35c.vcf -ploidy 1 -O KH35c.vcf -stand-call-conf 30 > genotype_gVCFs.out &
```

## Rename samples in vcf (to make it easier downstream)

```{bash bcftools, eval = FALSE}
bcftools reheader -s, --samples VCFs.list KH35c.vcf -o KH35c_rn.vcf
bcftools query -l KH35c_rn.vcf
```
