---
title: "Summary stats on VCF files"
author: "Rebecca Batstone"
date: "`r format(Sys.Date())`"
output: github_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load packages

```{r Load packages, message=FALSE, warning=FALSE}
library(tidyverse)
library(vcfR)
library(ape)
library("UpSetR") ## upset graphs
library(cowplot)
library(magick)

#function
Which.names <- function(DF, value){
  ind <- which(DF==value, arr.ind=TRUE)
  paste(rownames(DF)[ind[1:nrow(ind)]],  colnames(DF)[ind[1:nrow(ind),2]], sep=',')
}
```

# VCF summary stats using vcfR

Use vcfR package to examine important alignment parameters

## Full 1330 SNPs

### Summary statistics of vcf 

First, on server, subset full vcf into the four genomic regions (chromosome, accessory plasmid A, pSymA, pSymB). 

```{r subset_vcf_full, engine = 'bash', eval=FALSE}
# on server:
$VCFTOOLS --vcf  KH35c_filt1.recode.vcf --chr CP021825.1 --recode --recode-INFO-all --out KH35c_full_chrom
## 236 SNPs kept

$VCFTOOLS --vcf  KH35c_filt1.recode.vcf --chr CP021826.1 --recode --recode-INFO-all --out KH35c_full_pAccA
## 6 SNPs kept

$VCFTOOLS --vcf  KH35c_filt1.recode.vcf --chr CP021827.1 --recode --recode-INFO-all --out KH35c_full_pSymA
## 565 SNPs kept

$VCFTOOLS --vcf  KH35c_filt1.recode.vcf --chr CP021828.1 --recode --recode-INFO-all --out KH35c_full_pSymB
## 523 SNPs kept

## zipped on server, scp'ed to my comp. 
```

## Construct the chrom objects

```{r vcfR_full}
# reference file
ref <- ape::read.dna("./Data/GCA_002197105.1_ASM219710v1_genomic.fna", 
                     format = "fasta")

# chrom
ref_chrom <- ref[ grep( "CP021825.1 ", names(ref) ) ]
names(ref_chrom) <- "Chromosome"
ref_chrom <- as.matrix(ref_chrom)

# pAcc
ref_pAcc <- ref[ grep( "CP021826.1 ", names(ref) ) ]
names(ref_pAcc) <- "Accessory_plasmid"
ref_pAcc <- as.matrix(ref_pAcc)

# pSymA
ref_pSymA <- ref[ grep( "CP021827.1 ", names(ref) ) ]
names(ref_pSymA) <- "Symbiosis_plasmid_A"
ref_pSymA <- as.matrix(ref_pSymA)

# pSymB
ref_pSymB <- ref[ grep( "CP021828.1 ", names(ref) ) ]
names(ref_pSymB) <- "Symbiosis_plasmid_B"
ref_pSymB <- as.matrix(ref_pSymB)

# annotation file
gff <- read.table("./Data/GCA_002197105.1_ASM219710v1_genomic.gff", 
                  sep="\t", quote="")

gff_chrom <- gff[grep("NZ_CP021825.1", gff[,1]),]
gff_pAcc <- gff[grep("NZ_CP021826.1", gff[,1]),]
gff_pSymA <- gff[grep("NZ_CP021827.1", gff[,1]),]
gff_pSymB <- gff[grep("NZ_CP021828.1", gff[,1]),]

# vcf's
vcf_chrom <- read.vcfR("./Data/KH35c_full_regions/KH35c_full_chrom.recode.vcf")
vcf_pAcc <- read.vcfR("./Data/KH35c_full_regions/KH35c_full_pAccA.recode.vcf")
vcf_pSymA <- read.vcfR("./Data/KH35c_full_regions/KH35c_full_pSymA.recode.vcf")
vcf_pSymB <- read.vcfR("./Data/KH35c_full_regions/KH35c_full_pSymB.recode.vcf")

# create chrom objects
chrom <- create.chromR(name='Chromosome', vcf=vcf_chrom, seq=ref_chrom, ann=gff_chrom)
chrom_pAcc <- create.chromR(name='Accessory_plasmid', vcf=vcf_pAcc, seq=ref_pAcc, ann=gff_pAcc)
chrom_pSymA <- create.chromR(name='Symbiosis_plasmid_A', vcf=vcf_pSymA, seq=ref_pSymA, ann=gff_pSymA)
chrom_pSymB <- create.chromR(name='Symbiosis_plasmid_B', vcf=vcf_pSymB, seq=ref_pSymB, ann=gff_pSymB)

chrom <- proc.chromR(chrom, verbose=FALSE)
chrom_pAcc <- proc.chromR(chrom_pAcc, verbose=FALSE)
chrom_pSymA <- proc.chromR(chrom_pSymA, verbose=FALSE)
chrom_pSymB <- proc.chromR(chrom_pSymB, verbose=FALSE)

chromoqc(chrom, dp.alpha=20)
chromoqc(chrom_pAcc, dp.alpha=20)
chromoqc(chrom_pSymA, dp.alpha=20)
chromoqc(chrom_pSymB, dp.alpha=20)

plot(chrom)
plot(chrom_pAcc)
plot(chrom_pSymA)
plot(chrom_pSymB)

# Summarize per sample

# extract genotype
dp_chrom <- extract.gt(chrom, element="DP", as.numeric=TRUE)
rownames(dp_chrom) <- 1:nrow(dp_chrom)
head(dp_chrom)
heatmap.bp(dp_chrom)

dp_pAcc <- extract.gt(chrom_pAcc, element="DP", as.numeric=TRUE)
rownames(dp_pAcc) <- 1:nrow(dp_pAcc)
head(dp_pAcc)
heatmap.bp(dp_pAcc)

dp_pSymA <- extract.gt(chrom_pSymA, element="DP", as.numeric=TRUE)
rownames(dp_pSymA) <- 1:nrow(dp_pSymA)
head(dp_pSymA)
heatmap.bp(dp_pSymA)

dp_pSymB <- extract.gt(chrom_pSymB, element="DP", as.numeric=TRUE)
rownames(dp_pSymB) <- 1:nrow(dp_pSymB)
head(dp_pSymB)
heatmap.bp(dp_pSymB)

# visualizing read depth per sample, boxplot
par(mar=c(8,4,1,1))
#boxplot(dp, las=3, col=c("#C0C0C0", "#808080"), ylab="Depth", log='y', las=2)
boxplot(dp_chrom, las=3, col=c("#C0C0C0", "#808080"), ylab="Depth", las=2)
abline(h=seq(0,1e4, by=100), col="#C0C0C088")

par(mar=c(8,4,1,1))
#boxplot(dp, las=3, col=c("#C0C0C0", "#808080"), ylab="Depth", log='y', las=2)
boxplot(dp_pAcc, las=3, col=c("#C0C0C0", "#808080"), ylab="Depth", las=2)
abline(h=seq(0,1e4, by=100), col="#C0C0C088")

par(mar=c(8,4,1,1))
#boxplot(dp, las=3, col=c("#C0C0C0", "#808080"), ylab="Depth", log='y', las=2)
boxplot(dp_pSymA, las=3, col=c("#C0C0C0", "#808080"), ylab="Depth", las=2)
abline(h=seq(0,1e4, by=100), col="#C0C0C088")

par(mar=c(8,4,1,1))
#boxplot(dp, las=3, col=c("#C0C0C0", "#808080"), ylab="Depth", log='y', las=2)
boxplot(dp_pSymB, las=3, col=c("#C0C0C0", "#808080"), ylab="Depth", las=2)
abline(h=seq(0,1e4, by=100), col="#C0C0C088")
```

## 363 SNPs used in GWAS

### Summary statistics of vcf 

First, on server, subset full vcf into the four genomic regions (chromosome, accessory plasmid A, pSymA, pSymB). 

```{r subset_vcf_gwas, engine = 'bash', eval=FALSE}
# on server:
$VCFTOOLS --vcf  KH35c_GEMMA.recode.vcf --chr CP021825.1 --recode --recode-INFO-all --out KH35c_gwas_chrom
## 34 SNPs kept

$VCFTOOLS --vcf  KH35c_GEMMA.recode.vcf --chr CP021826.1 --recode --recode-INFO-all --out KH35c_gwas_pAccA
## 2 SNPs kept

$VCFTOOLS --vcf  KH35c_GEMMA.recode.vcf --chr CP021827.1 --recode --recode-INFO-all --out KH35c_gwas_pSymA
## 184 SNPs kept

$VCFTOOLS --vcf  KH35c_GEMMA.recode.vcf --chr CP021828.1 --recode --recode-INFO-all --out KH35c_gwas_pSymB
## 143 SNPs kept

## zipped on server, scp'ed to my comp. 
```

## Construct the chrom objects

```{r vcfR_gwas}
# vcf's
vcf_chrom <- read.vcfR("./Data/KH35c_gwas_regions/KH35c_gwas_chrom.recode.vcf")
vcf_pAcc <- read.vcfR("./Data/KH35c_gwas_regions/KH35c_gwas_pAccA.recode.vcf")
vcf_pSymA <- read.vcfR("./Data/KH35c_gwas_regions/KH35c_gwas_pSymA.recode.vcf")
vcf_pSymB <- read.vcfR("./Data/KH35c_gwas_regions/KH35c_gwas_pSymB.recode.vcf")

# create chrom objects
chrom <- create.chromR(name='Chromosome', vcf=vcf_chrom, seq=ref_chrom, ann=gff_chrom)
chrom_pAcc <- create.chromR(name='Accessory_plasmid', vcf=vcf_pAcc, seq=ref_pAcc, ann=gff_pAcc)
chrom_pSymA <- create.chromR(name='Symbiosis_plasmid_A', vcf=vcf_pSymA, seq=ref_pSymA, ann=gff_pSymA)
chrom_pSymB <- create.chromR(name='Symbiosis_plasmid_B', vcf=vcf_pSymB, seq=ref_pSymB, ann=gff_pSymB)

chrom <- proc.chromR(chrom, verbose=FALSE, win.size = 100)
chrom_pAcc <- proc.chromR(chrom_pAcc, verbose=FALSE, win.size = 100)
chrom_pSymA <- proc.chromR(chrom_pSymA, verbose=FALSE, win.size = 100)
chrom_pSymB <- proc.chromR(chrom_pSymB, verbose=FALSE, win.size = 100)

chromoqc(chrom, dp.alpha=20)
chromoqc(chrom_pAcc, dp.alpha=20)
chromoqc(chrom_pSymA, dp.alpha=20)
chromoqc(chrom_pSymB, dp.alpha=20)

plot(chrom)
plot(chrom_pAcc)
plot(chrom_pSymA)
plot(chrom_pSymB)

# Summarize per sample

# extract genotype
dp_chrom <- extract.gt(chrom, element="DP", as.numeric=TRUE)
rownames(dp_chrom) <- 1:nrow(dp_chrom)
head(dp_chrom)
heatmap.bp(dp_chrom)

dp_pAcc <- extract.gt(chrom_pAcc, element="DP", as.numeric=TRUE)
rownames(dp_pAcc) <- 1:nrow(dp_pAcc)
head(dp_pAcc)
heatmap.bp(dp_pAcc)

dp_pSymA <- extract.gt(chrom_pSymA, element="DP", as.numeric=TRUE)
rownames(dp_pSymA) <- 1:nrow(dp_pSymA)
head(dp_pSymA)
heatmap.bp(dp_pSymA)

dp_pSymB <- extract.gt(chrom_pSymB, element="DP", as.numeric=TRUE)
rownames(dp_pSymB) <- 1:nrow(dp_pSymB)
head(dp_pSymB)
heatmap.bp(dp_pSymB)

# visualizing read depth per sample, boxplot
par(mar=c(8,4,1,1))
#boxplot(dp, las=3, col=c("#C0C0C0", "#808080"), ylab="Depth", log='y', las=2)
boxplot(dp_chrom, las=3, col=c("#C0C0C0", "#808080"), ylab="Depth", las=2)
abline(h=seq(0,1e4, by=100), col="#C0C0C088")

par(mar=c(8,4,1,1))
#boxplot(dp, las=3, col=c("#C0C0C0", "#808080"), ylab="Depth", log='y', las=2)
boxplot(dp_pAcc, las=3, col=c("#C0C0C0", "#808080"), ylab="Depth", las=2)
abline(h=seq(0,1e4, by=100), col="#C0C0C088")

par(mar=c(8,4,1,1))
#boxplot(dp, las=3, col=c("#C0C0C0", "#808080"), ylab="Depth", log='y', las=2)
boxplot(dp_pSymA, las=3, col=c("#C0C0C0", "#808080"), ylab="Depth", las=2)
abline(h=seq(0,1e4, by=100), col="#C0C0C088")

par(mar=c(8,4,1,1))
#boxplot(dp, las=3, col=c("#C0C0C0", "#808080"), ylab="Depth", log='y', las=2)
boxplot(dp_pSymB, las=3, col=c("#C0C0C0", "#808080"), ylab="Depth", las=2)
abline(h=seq(0,1e4, by=100), col="#C0C0C088")
```

# Add relevant info to vcf summary files

Two files:
  * filt1_1330.recode.vcf (1330 sites)
  * GEMMA_363.recode.vcf (363 sites)
  
Info to add:
  * SnpEff annotations
  * NCBI annotations (GEMMA_363 only)
  * Minor allele (maf)
  * Evolutionary origin
  * maf for each line of origin

## Annotate VCFs using SnpEff (build database)

First, need to build database in SnpEff:

```{r build_snpEff, engine= 'bash', eval = FALSE}
wget http://sourceforge.net/projects/snpeff/files/snpEff_latest_core.zip
# unzip
cd SnpEff

# check to see what genomes are already in the database:
java -jar snpEff.jar databases | grep -i Sinorhizobium_meliloti
# found 1021, but not KH35c. Had to build my own...

# build database using GFF (GTF preferred, but not available for KH35c)
mkdir path/to/snpEff/data/Sinorhizobium_meliloti_kh35c
cd path/to/snpEff/data/Sinorhizobium_meliloti_kh35c
wget ftp://ftp.ncbi.nlm.nih.gov/genomes/genbank/bacteria/Sinorhizobium_meliloti/latest_assembly_versions/GCA_002197105.1_ASM219710v1/GCA_002197105.1_ASM219710v1_genomic.gff.gz
mv GCA_002197105.1_ASM219710v1_genomic.gff.gz genes.gff.gz 
# note: renaming to genes.gff.gz is essential

# Get the genome
cd /path/to/snpEff/data/genomes
wget ftp://ftp.ncbi.nlm.nih.gov/genomes/genbank/bacteria/Sinorhizobium_meliloti/latest_assembly_versions/GCA_002197105.1_ASM219710v1/GCA_002197105.1_ASM219710v1_genomic.fna.gz
mv GCA_002197105.1_ASM219710v1_genomic.fna.gz Sinorhizobium_meliloti_kh35c.fa.gz
# needs to match name of directory in data/, fa and not fna!

# Add the new genome to the config file
cd path/to/snpEff/
nano snpEffect.config 
# Sinorhizobium genome, version KH35c
Sinorhizobium_meliloti_kh35c.genome : Sinorhizobium_meliloti_kh35c

#create database
java -jar snpEff.jar build -gff3 -v Sinorhizobium_meliloti_kh35c
```

## Use SnpEff to get annotations (filt1_1330)

```{r vcf_annotate_snpEff, engine= 'bash', eval = FALSE}
# /gran1/rebecca.batstone/align_KH35c_08Mar2018/processed_bam_samples:
java -Xmx4g -jar $SNPEFF Sinorhizobium_meliloti_kh35c -v -s 'snpEff_filt1_1330.html' -ud 0 filt1_1330.recode.vcf > filt1_1330.ann.vcf

# scp'ed summary files over: 
scp rebecca.batstone@grandiflora.eeb.utoronto.ca:/gran1/rebecca.batstone/align_KH35c_08Mar2018/processed_bam_samples/snpEff_filt1_1330* /c/Users/$USERNAME/Dropbox/'Evolution experiment 2016'/ens_evo_exp/VCF_summary_files/

# summarize vcf to table
java -jar $GATK3 -R /gran1/rebecca.batstone/align_KH35c_08Mar2018/processed_bam_samples/GCA_002197105.1_ASM219710v1_genomic.fna -T VariantsToTable -V filt1_1330.ann.vcf -F CHROM -F POS -F REF -F ALT -F QUAL -F AF -F ANN -F DP -GF GT -o filt1_1330.ann.vcf.table

# scp'ed to VCF summary folder
scp rebecca.batstone@grandiflora.eeb.utoronto.ca:/gran1/rebecca.batstone/align_KH35c_08Mar2018/processed_bam_samples/filt1_1330.ann.vcf.table /c/Users/$USERNAME/Dropbox/'Evolution experiment 2016'/ens_evo_exp/VCF_summary_files/
```

## Add information to VCF summary file (filt1_1330)

```{r summary_1330}
# Import VCF info file, merge
filt1_vars <- read.table("./Data/filt1_1330.ann.vcf.table",header=T,sep="\t")
## rename column POS to ps for downstream merging
colnames(filt1_vars)[colnames(filt1_vars)=="POS"] <- "ps"

# designate REF or ALT as minor or major allele
filt1_vars$minor <- filt1_vars$AF < 0.5
filt1_vars$min_allele <- ifelse(filt1_vars$minor == "TRUE", as.character(filt1_vars$ALT),
                                  as.character(filt1_vars$REF))

# add in evolutionary origin:
## First, need to make sure all variants are encoded as characters
rownames(filt1_vars) <- filt1_vars[,2] ## name rows by position
cols.to.character <- 9:59
filt1_vars[cols.to.character] <- sapply(filt1_vars[cols.to.character],as.character)
## Then, specify cols corresponding to each host line:
cols.267 <- 9:16
filt1_vars$mac_267 <- (rowSums(filt1_vars[,cols.267] == filt1_vars$min_allele)) ## number of samples for 267
filt1_vars$maf_267 <- filt1_vars$mac_267/8 
cols.270 <- 17:28
filt1_vars$mac_270 <- (rowSums(filt1_vars[,cols.270] == filt1_vars$min_allele)) ## number of samples for 270
filt1_vars$maf_270 <- filt1_vars$mac_270/12
cols.276 <- 29:31
filt1_vars$mac_276 <- (rowSums(filt1_vars[,cols.276] == filt1_vars$min_allele)) ## number of samples for 276
filt1_vars$maf_276 <- filt1_vars$mac_276/3 
cols.279 <- 32:42
filt1_vars$mac_279 <- (rowSums(filt1_vars[,cols.279] == filt1_vars$min_allele)) ## number of samples for 279
filt1_vars$maf_279 <- filt1_vars$mac_279/11 
cols.313 <- 43:48
filt1_vars$mac_313 <- (rowSums(filt1_vars[,cols.313] == filt1_vars$min_allele)) ## number of samples for 313
filt1_vars$maf_313 <- filt1_vars$mac_313/6 
cols.anc <- 49:57
filt1_vars$mac_anc <- (rowSums(filt1_vars[,cols.anc] == filt1_vars$min_allele)) ## number of samples for anc
filt1_vars$maf_anc <- filt1_vars$mac_anc/9 
# overall mac
filt1_vars$mac <-  filt1_vars$mac_267 + filt1_vars$mac_270 + filt1_vars$mac_276 + 
  filt1_vars$mac_279 + filt1_vars$mac_313 + filt1_vars$mac_anc ## total number of minor alleles
# set to true if maf > 0
origin <- replace(filt1_vars[,c(61,63,65,67,69,71)], filt1_vars[,c(61,63,65,67,69,71)] > 0,"TRUE")
# change names to shorter version
names(origin) <- c("L267","L270", "L276", "L279", "L313", "anc")
# create a function to pull out all sig (=TRUE) vars, then paste the position and traits (col names)
# get it into the right format

origin_vars <- sort(Which.names(origin, value=TRUE))
origin_vars_df <- as.data.frame(origin_vars)
origin_vars.s <- separate(data = origin_vars_df, col = origin_vars, into = c("ps", "origin"), sep = "\\,")
origin_vars.a <- aggregate(origin ~ ps, data=origin_vars.s, toString)
# merge with variant info file:
filt1_vars.o <- merge(filt1_vars, origin_vars.a, all.x = TRUE, by = "ps")
# drop unwanted cols
filt1_vars.o <- filt1_vars.o[ , -c(9:58)]

# add in whether de novo or not
filt1_vars.o$Novel <- filt1_vars.o$maf_anc
filt1_vars.o$Novel <-  replace(filt1_vars.o$Novel, filt1_vars.o$Novel > 0,"No")
filt1_vars.o$Novel <-  replace(filt1_vars.o$Novel, filt1_vars.o$Novel == 0,"Yes")

# extract gene, impact, function from snpEff annotation
filt1_vars.o$ANN <- as.character(filt1_vars.o$ANN)
filt1_vars.a <- separate(data = filt1_vars.o, col = ANN, into = c("allele", "type","impact","gene"), sep = "\\|")
## drop allele
filt1_vars.a <- filt1_vars.a[ , -which(names(filt1_vars.a) %in% c("allele"))]

# add in var type (SNP, del or ins)
filt1_vars.a$n_REF <- nchar(as.character(filt1_vars.a$REF))
filt1_vars.a$n_ALT <- nchar(as.character(filt1_vars.a$ALT))
filt1_vars.a$var_type <- ifelse(filt1_vars.a$n_REF == filt1_vars.a$n_ALT, "SNP", 
                                ifelse(filt1_vars.a$n_REF > filt1_vars.a$n_ALT, "DEL", "INS"))

filt1_vars.f <- filt1_vars.a[,c(1:10, 25:26, 29, 11:24)]

write.csv(filt1_vars.f, "./Output/filt1_1330_sum.csv", row.names = FALSE)
save(filt1_vars.f, file = "./Output/filt1_1330_sum.Rdata")
```

## Minor allele frequency spectra (1330)

```{r maf_spectra}
load(file = "./Output/filt1_1330_sum.Rdata") ## filt1_vars.f

filt1_vars.f_span <- filt1_vars.f %>%
  filter(!is.na(origin)) %>%
     group_by(Novel, AC_bin=cut(mac, breaks= seq(0, 24, by = 3))) %>%
     summarise(count = n()) %>%
     arrange(as.numeric(AC_bin)) %>%
     ungroup(.) %>%
     group_by(Novel) %>%
     mutate(total = sum(count), prop = count/total)

level_key <- c('(0,3]' = 3,  '(3,6]' = 6, '(6,9]' = 9, '(9,12]' = 12,
               '(12,15]' = 15, '(15,18]' = 18, '(18,21]' = 21, '(21,24]' = 24)
filt1_vars.f_span$mac <- recode_factor(filt1_vars.f_span$AC_bin, !!!level_key)

png(filename="./AFS_allvars.png", units="px", width=797, height=557)
     
ggplot(filt1_vars.f_span, aes(x = mac, y = prop, fill = Novel)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme_bw() +
  xlab("Minor allele count") +
  ylab("Proportion of loci") +
  theme(axis.text.y = element_text(size=12), 
        legend.position=c(0.8,0.8), 
        legend.background = element_blank(),
        legend.key = element_rect(colour = "transparent", fill = "transparent"),
        axis.title.y = element_text(size=16), 
        axis.title.x = element_text(size=16),
        axis.text.x = element_text(size=12), 
        plot.title = element_text(size=16, face = "bold"),
        panel.background = element_rect(fill="transparent"),
        strip.background = element_rect(fill="transparent"),
        strip.text = element_text(size = 12, face = "bold"),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()
        )
  
dev.off()

## do the same, but split for novel isolates within each host line
filt1_vars.f$group <-ifelse(grepl("anc", filt1_vars.f$origin, fixed = "FALSE"), 
                            "standing",
                            ifelse(filt1_vars.f$origin == "L270", "unique_270",
                                   ifelse(filt1_vars.f$origin == "L276", "unique_276",
                                          ifelse(filt1_vars.f$origin == "L279", "unique_279",
                                                 ifelse(filt1_vars.f$origin == "L313", "unique_313",
                                                        ifelse(filt1_vars.f$origin == "L267",
                                                               "unique_267","exclude"))))))
```

## Use SnpEff to get annotations (GEMMA_363) 

```{r vcf_summary, engine= 'bash', eval = FALSE}
# From the GEMMA directory
java -Xmx4g -jar $SNPEFF Sinorhizobium_meliloti_kh35c -v -s 'snpEff_GEMMA_363.html' -ud 0  GEMMA_363.recode.vcf > GEMMA_363.ann.vcf

# scp'ed summary files over: 
scp rebecca.batstone@grandiflora.eeb.utoronto.ca:/gran1/rebecca.batstone/align_KH35c_08Mar2018/processed_bam_samples/snpEff_GEMMA_363* /c/Users/$USERNAME/Dropbox/'Evolution experiment 2016'/ens_evo_exp/VCF_summary_files/

# summarize vcf to table
java -jar $GATK3 -R /gran1/rebecca.batstone/align_KH35c_08Mar2018/processed_bam_samples/GCA_002197105.1_ASM219710v1_genomic.fna -T VariantsToTable -V GEMMA_363.ann.vcf -F CHROM -F POS -F REF -F ALT -F QUAL -F AF -F ANN -F DP -GF GT -o GEMMA_363.ann.vcf.table

# scp'ed to VCF summary folder
scp rebecca.batstone@grandiflora.eeb.utoronto.ca:/gran1/rebecca.batstone/GEMMA/GEMMA_363.ann.vcf.table /c/Users/$USERNAME/Dropbox/'Evolution experiment 2016'/ens_evo_exp/VCF_summary_files/
```

## Use bedtools to get annotations (GEMMA_363) 

Note: there is a diff btw GCA (genbank) and GCF (Refseq). All calls were made on GCA.

```{r bedtools, engine='bash', eval = FALSE}
# Gene names for each variant
$BEDTOOLS intersect -loj -a GEMMA_363.ann.vcf -b GCA_002197105.1_ASM219710v1_genomic.gff -wb | \
  cut -f 1,2,61,67 > GEMMA_363.genes.txt 
  ## extract region, ps, field, and info
  
# scp'ed to VCF summary folder
scp rebecca.batstone@grandiflora.eeb.utoronto.ca:/gran1/rebecca.batstone/GEMMA/GEMMA_363.genes.txt /c/Users/IGB/Dropbox/'Evolution experiment 2016'/ens_evo_exp/VCF_summary_files/
```

## Add information to VCF summary file (GEMMA_363)

```{r merge_sum_363}
# Import VCF info file, merge
GEMMA_vars <- read.table("./Data/GEMMA_363.ann.vcf.table",header=T,sep="\t")
## rename column POS to ps for downstream merging
colnames(GEMMA_vars)[colnames(GEMMA_vars)=="POS"] <- "ps"

# designate REF or ALT as minor or major allele
GEMMA_vars$minor <- GEMMA_vars$AF < 0.5
GEMMA_vars$min_allele <- ifelse(GEMMA_vars$minor == "TRUE", as.character(GEMMA_vars$ALT),
                                  as.character(GEMMA_vars$REF))

# add in evolutionary origin:
## First, need to make sure all variants are encoded as characters
rownames(GEMMA_vars) <- GEMMA_vars[,2] ## name rows by position
cols.to.character <- 9:59
GEMMA_vars[cols.to.character] <- sapply(GEMMA_vars[cols.to.character],as.character)
## Then, specify cols corresponding to each host line:
cols.267 <- 9:16
GEMMA_vars$maf_267 <- (rowSums(GEMMA_vars[,cols.267] == GEMMA_vars$min_allele))/8 ## number of samples for 267
cols.270 <- 17:28
GEMMA_vars$maf_270 <- (rowSums(GEMMA_vars[,cols.270] == GEMMA_vars$min_allele))/12 ## number of samples for 270
cols.276 <- 29:31
GEMMA_vars$maf_276 <- (rowSums(GEMMA_vars[,cols.276] == GEMMA_vars$min_allele))/3 ## number of samples for 276
cols.279 <- 32:42
GEMMA_vars$maf_279 <- (rowSums(GEMMA_vars[,cols.279] == GEMMA_vars$min_allele))/11 ## number of samples for 279
cols.313 <- 43:48
GEMMA_vars$maf_313 <- (rowSums(GEMMA_vars[,cols.313] == GEMMA_vars$min_allele))/6 ## number of samples for 313
cols.anc <- 49:57
GEMMA_vars$maf_anc <- (rowSums(GEMMA_vars[,cols.anc] == GEMMA_vars$min_allele))/9 ## number of samples for anc
# calculate total minor allele frequency
cols.all <- 9:57
GEMMA_vars$maf <- (rowSums(GEMMA_vars[,cols.all] == GEMMA_vars$min_allele))/49 ## number of samples
# set to true if maf > 0
origin <- replace(GEMMA_vars[,c(65,61,62,63,64,60)], 
                  GEMMA_vars[,c(65,61,62,63,64,60)] > 0,"TRUE")
# change names to shorter version
names(origin) <- c("anc", "270", "276", "279", "313", "267")
# get it into the right format
origin_vars <- Which.names(origin, value=TRUE)
origin_vars_df <- as.data.frame(origin_vars)
origin_vars.s <- separate(data = origin_vars_df, 
                          col = origin_vars, into = c("ps", "origin"), sep = "\\,")
origin_vars.a <- aggregate(origin ~ ps, data=origin_vars.s, toString)
# merge with variant info file:
GEMMA_vars.o <- merge(GEMMA_vars, origin_vars.a, all.x = TRUE, by = "ps")
# drop unwanted cols
GEMMA_vars.o <- GEMMA_vars.o[ , -c(9:58)]

# add in whether de novo or not
GEMMA_vars.o$Novel <- GEMMA_vars.o$maf_anc
GEMMA_vars.o$Novel <-  replace(GEMMA_vars.o$Novel, GEMMA_vars.o$Novel > 0,"No")
GEMMA_vars.o$Novel <-  replace(GEMMA_vars.o$Novel, GEMMA_vars.o$Novel == 0,"Yes")

# extract gene, impact, function from snpEff annotation
GEMMA_vars.o$ANN <- as.character(GEMMA_vars.o$ANN)
GEMMA_vars.a <- separate(data = GEMMA_vars.o, col = ANN, into = c("allele", "type","impact","gene"), sep = "\\|")
## drop allele
GEMMA_vars.a <- GEMMA_vars.a[ , -which(names(GEMMA_vars.a) %in% c("allele"))]

# add in gene products
func <- read.delim("./Data/GEMMA_363.genes.txt", header=FALSE) ## note, tsv only imported half the file
## rename column POS to ps to merge
colnames(func) <- c("CHROM","ps","field","info")
# filter out coding sequences
func.f <- func %>%
  filter(field == "CDS")
# extract gene function
func.f$ncbi_func <- sub(".*product= *(.*?) *;.*", "\\1", func.f$info)
func.f$RefSeq_ID <- sub(".*RefSeq: *(.*?) *;.*", "\\1", func.f$info)
func.f$protein_ID <- sub(".*protein_id= *(.*?) *;.*", "\\1", func.f$info)
func.f$protein_ID <- gsub("[ID=]", NA, func.f$protein_ID) ## replace long string with NA
func.f$gene_ID <- sub(".*Parent= *(.*?) *;.*", "\\1", func.f$info)
# merge info
GEMMA_vars.p <- GEMMA_vars.a
GEMMA_vars.p$ncbi_func <- func.f$ncbi_func[match(GEMMA_vars.p$ps, func.f$ps)]
# rename NA in ncbi function
GEMMA_vars.p$ncbi_func[is.na(GEMMA_vars.p$ncbi_func)] <- "intergenic"
GEMMA_vars.p$RefSeq_ID <- func.f$RefSeq_ID[match(GEMMA_vars.p$ps, func.f$ps)]
GEMMA_vars.p$protein_ID <- func.f$protein_ID[match(GEMMA_vars.p$ps, func.f$ps)]
GEMMA_vars.p$gene_ID <- func.f$gene_ID[match(GEMMA_vars.p$ps, func.f$ps)]

# add in var type (SNP, del or ins)
GEMMA_vars.p$n_REF <- nchar(as.character(GEMMA_vars.p$REF))
GEMMA_vars.p$n_ALT <- nchar(as.character(GEMMA_vars.p$ALT))
GEMMA_vars.p$var_type <- ifelse(GEMMA_vars.p$n_REF == GEMMA_vars.p$n_ALT, "SNP", 
                                ifelse(GEMMA_vars.p$n_REF > GEMMA_vars.p$n_ALT, "DEL", "INS"))

# create final version of GEMMA_vars
GEMMA_vars.f <- GEMMA_vars.p[,c(1:10, 19:24, 27, 11:18)]

write.csv(GEMMA_vars.f, "./Output/GEMMA_363_sum.csv", row.names = FALSE)
save(GEMMA_vars.f, file = "./Output/GEMMA_363_sum.Rdata")
```

## Figures

### Figure S5: UpSet R graphs

```{r upset, warning = FALSE}
# add variant pres/abs for each line of origin
upset_1330 <- replace(filt1_vars[,60:65], filt1_vars[,60:65] > 0,1)
# change names to shorter version
names(upset_1330) <- c("L267","L270", "L276", "L279", "L313", "ancestral")

# add function to color bars by whether they are novel or not
Myfunc <- function(row, anc) {
    data <- (row["ancestral"] == anc)
}

tiff('./Data/upset1330.tif', width = 10, height = 6, units = "in", res = 300)

upset(upset_1330, order.by = "freq", 
      nintersects = 28,
      #cutoff = 5,
      text.scale = c(2, 2, 1.5, 1.5, 2, 1.5),
      query.legend = "none",
      main.bar.color = "black", 
      mainbar.y.label = "Shared variants (no.)", 
      sets.x.label = "Total variants (no.)",
      sets = c("L267","L270", "L276", "L279", "L313", "ancestral"),
      keep.order = TRUE,
      sets.bar.color = c("black","black", "black", 
                         "black", "black", "red"),
      queries = list(list(query = Myfunc, 
                                    params = list(1), 
                                    color = "red", 
                                    active = T,
                          query.name = "Standing genetic variation"),
                     list(query = Myfunc, 
                                    params = list(0), 
                                    color = "black", 
                                    active = T,
                          query.name = "De novo")))

dev.off()

## for 363 var
# add variant pres/abs for each line of origin
upset_363 <- replace(GEMMA_vars[,60:65], GEMMA_vars[,60:65] > 0,1)
# change names to shorter version
names(upset_363) <- c("L267","L270", "L276", "L279", "L313", "ancestral")

tiff('./Data/upset363.tif', width = 10, height = 6, units = "in", res = 300)

upset(upset_363, order.by = "freq", 
      nintersects = 19,
      #cutoff = 5,
      text.scale = c(2, 2, 1.5, 1.5, 2, 1.5),
      query.legend = "none",
      main.bar.color = "black", 
      mainbar.y.label = "Shared variants (no.)", 
      sets.x.label = "Total variants (no.)",
      sets = c("L267","L270", "L276", "L279", "L313", "ancestral"),
      keep.order = TRUE,
      sets.bar.color = c("black","black", "black", 
                         "black", "black", "red"),
      queries = list(list(query = Myfunc, 
                                    params = list(1), 
                                    color = "red", 
                                    active = T,
                          query.name = "Standing genetic variation"),
                     list(query = Myfunc, 
                                    params = list(0), 
                                    color = "black", 
                                    active = T,
                          query.name = "De novo")))

dev.off()

# summary info
## how many novel SNPs?
filt1_vars.a %>%
  na.omit() %>%
  group_by(Novel) %>%
  summarize(count = n()) %>%
  mutate(perc = (count/sum(count))*100)

GEMMA_vars.o %>%
  na.omit() %>%
  group_by(Novel) %>%
  summarize(count = n()) %>%
  mutate(perc = (count/sum(count))*100)

## Save combined plots

upset_1330 <- ggdraw() + draw_image("./Data/upset1330.tif", scale = 1)
var_info_1330 <- ggdraw() + draw_image("./Data/filt1_1330_info.tif", scale = 1)

var1330_inset <- ggdraw(upset_1330 + theme_half_open(12)) +
  draw_plot(var_info_1330, .5, .45, .5, .5) +
  draw_plot_label(
    c("A", "B"),
    c(0.1, 0.5),
    c(0.95, 0.95),
    size = 12)

upset_363 <- ggdraw() + draw_image("./Data/upset363.tif", scale = 1)
var_info_363 <- ggdraw() + draw_image("./Data/GEMMA_363_info.tif", scale = 1)

var363_inset <- ggdraw(upset_363 + theme_half_open(12)) +
  draw_plot(var_info_363, .5, .5, .5, .5) +
  draw_plot_label(
    c("C", "D"),
    c(0.1, 0.5),
    c(0.95, 0.95),
    size = 12)

# combine
fig_upset <- plot_grid(var1330_inset, var363_inset, 
          ncol = 1,
          nrow = 2,
          align = "v",
          labels = NULL)

save_plot("./Figures/FigureS5.pdf", fig_upset,
          ncol = 1, # we're saving a grid plot of 2 columns
          nrow = 2, # and 3 rows
          # each individual subplot should have an aspect ratio of 1.3
          base_aspect_ratio = 1.3
          )

```